### PR TITLE
allow differently colored codes

### PR DESCRIPTION
**Pull request by flipace**
this allows detection of codes which have different color/contrast than black on white.
achieved by sending a negative version of every second frame to the google vision library

### CODE REVIEW

File: src/ios/CameraViewController.m [hunk 1]
```diff
@@ -37,9 +36,88 @@ @interface CameraViewController ()<AVCaptureVideoDataOutputSampleBufferDelegate>
 
 @end
 
+@implementation UIImage (NegativeImage)
```
Comment (flipace) [inline src/ios/CameraViewController.m:40]:
```suggestion
// muchas gracias to: https://stackoverflow.com/a/6672628/1856221
@implementation UIImage (NegativeImage)
```

File: src/android/src/gmvScanner/ui/camera/CameraSource.java [hunk 2]
```diff
@@ -1135,6 +1140,17 @@ void setNextFrame(byte[] data, Camera camera) {
                 // idea of the timing of frames received and when frames were dropped along the way.
                 mPendingTimeMillis = SystemClock.elapsedRealtime() - mStartTimeMillis;
                 mPendingFrameId++;
+             
+
+                if (!isInverted){
+                    for (int y = 0; y < data.length; y++){
+                        data[y] = (byte) ~data[y];
+                    }
+                    isInverted = true;
```
Comment (prk3) [inline src/android/src/gmvScanner/ui/camera/CameraSource.java]: Safer option - `isInverted = ~isInverted;` below all other code.

File: src/ios/CameraViewController.m [hunk 3]
```diff
@@ -178,6 +261,14 @@ - (void)captureOutput:(AVCaptureOutput *)captureOutput
     CGRect cropRect = CGRectMake(imageWidth/2 - actualFrameWidth/2, imageHeight/2 - actualFrameHeight/2, actualFrameWidth, actualFrameHeight);
     //Crop image
     croppedImg = [self croppIngimageByImageName:portraitImage toRect:cropRect];
+
+    // invert image colors
+    if(!_isInverted) {
+        croppedImg = [croppedImg negativeImage];
+        _isInverted = true;
```
Comment (prk3) [inline src/ios/CameraViewController.m]: You know the deal.

File: src/ios/CameraViewController.m [hunk 4]
```diff
@@ -37,9 +36,89 @@ @interface CameraViewController ()<AVCaptureVideoDataOutputSampleBufferDelegate>
...
+{
+    // get width and height as integers, since we'll be using them as
+    // array subscripts, etc, and this'll save a whole lot of casting
+    CGSize size = self.size;
+    int width = size.width;
+    int height = size.height;
+    
+    // Create a suitable RGB+alpha bitmap context in BGRA colour space
+    CGColorSpaceRef colourSpace = CGColorSpaceCreateDeviceRGB();
+    unsigned char *memoryPool = (unsigned char *)calloc(width*height*4, 1);
```
Comment (prk3) [inline src/ios/CameraViewController.m:52]: UHG
what is this monstrosity of a language?

**Review by prk3 — CHANGES_REQUESTED** [review]
You must have had wonderful time writing in this weird language.
Other than boolean value inverting, it's looking magnificent.

File: src/ios/CameraViewController.m (hunk 3 above)
Comment (flipace) [reply src/ios/CameraViewController.m]:
```suggestion
        _isInverted = !_isInverted;
```

File: src/ios/CameraViewController.m (hunk 4 above)
Comment (flipace) [reply src/ios/CameraViewController.m:52]: Well, I don't know a single person that likes objective c. And I think I don't know a language which is so widespread (or even mandatory) and this ugly. 🙈🙉🙊

**Review by Southgarden116 — APPROVED** [review]

**Pull request merged**